scalyr_agent/builtin_monitors/journald_monitor.py

```diff
@@ -0,0 +1,258 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ------------------------------------------------------------------------
+# author:  Imron Alston <imron@scalyr.com>
+
+__author__ = 'imron@scalyr.com'
+
+import datetime
+import os
+import re
+import select
+from scalyr_agent import ScalyrMonitor, define_config_option
+from scalyr_agent.scalyr_monitor import BadMonitorConfiguration
+from scalyr_agent.monitor_utils import journald_checkpoints
+from scalyr_agent.monitor_utils.journald_checkpoints import Checkpoint
+import scalyr_agent.scalyr_logging as scalyr_logging
+
+try:
+    from systemd import journal
+except ImportError:
+    raise Exception('Python systemd library not installed.\n\nYou must install the systemd python library in order '
+                    'to use the journald monitor.\n\nThis can be done via package manager e.g.:\n\n'
+                    '  apt-get install python-systemd  (debian/ubuntu)\n'
+                    '  dnf install python-systemd  (CentOS/rhel/Fedora)\n\n'
+                    'or installed from source using pip e.g.\n\n'
+                    '  pip install systemd-python\n\n'
+                    'See here for more info: https://github.com/systemd/python-systemd/\n'
+                   )
+
+global_log = scalyr_logging.getLogger(__name__)
+__monitor__ = __name__
+
+define_config_option( __monitor__, 'journal_path',
+                     'Optional (defaults to /var/log/journal). Location on disk of the journald logs.',
+                     convert_to=str, default='/var/log/journal')
+
+define_config_option( __monitor__, 'journal_poll_timeout_milliseconds',
+                     'Optional (defaults to 0). The number of milliseconds to wait for data while polling the journal file. '
+                     'Note: you are better off setting the sample_interval monitor option if the timeout is longer than '
+                     'one second.',
+                     convert_to=int, default=0)
+
+define_config_option( __monitor__, 'journal_fields',
+                     'Optional dict containing a list of journal fields to include with each message, as well as a field name to map them to.\n'
+                     'Note: Not all fields need to exist in every message and only fields that exist will be included.\n'
+                     'Defaults to {:\n'
+                     '  "_SYSTEMD_UNIT": "unit"\n'
+                     '  "_PID": "pid"\n'
+                     '  "_MACHINE_ID": "machine_id"\n'
+                     '  "_BOOT_ID": "boot_id"\n'
+                     '  "_HOSTNAME": "hostname"\n'
+                     '  "_SOURCE_REALTIME_TIMESTAMP": timestamp"\n'
+                     '}\n',
+                     default=None
+                     )
+
+define_config_option( __monitor__, 'journal_matches',
+                      'Optional list containing match strings for filtering entries.  A match string consists of "FIELD=value" where FIELD '
```

Comment:
I believe you said we are relying on a standard journald filtering mechanism here.  Our documentation here should point to their documentation.  I'm not quite understanding how this works from the description here.